File: src/Utils/typeName.ts
import { ExpectationFailedTJSGError } from "../Error/Errors.js";
import type { RawType, RawTypeName } from "../Schema/RawType.js";

export function typeName(value: RawType): RawTypeName {
    if (value === null) {
        return "null";
    }

    const type = typeof value;

    if (type === "string" || type === "number" || type === "boolean") {
        return type;
    }

    if (Array.isArray(value)) {
        return "array";
    }

    if (type === "object") {
        return "object";
    }

    throw new ExpectationFailedTJSGError(`JavaScript type "typeof " can't be converted to JSON type name`);
}

File: internal/formatter/type_name.go
package formatter

import (
	"fmt"
)

// typeName maps a raw JSON value to its JSON Schema type name
// (src/Utils/typeName.ts).
func typeName(value any) string {
	switch value.(type) {
	case nil:
		return "null"
	case string:
		return "string"
	case float64, int:
		return "number"
	case bool:
		return "boolean"
	case []any:
		return "array"
	case map[string]any:
		return "object"
	}
	panic(fmt.Errorf("JavaScript type %T can't be converted to JSON type name", value))
}

// toEnumType unwraps the slice if it contains only one type name
// (src/TypeFormatter/EnumTypeFormatter.ts).
func toEnumType(names []string) any {
	if len(names) == 1 {
		return names[0]
	}
	return names
}
